### PR TITLE
ST: Asserts for logs

### PR DESCRIPTION
**Pull request by serrss**
### Type of change
- Enhancement / new feature

### Description

Added method to search errors(exceptions) in logs for any type of resources.
Added asserts to check that after deleting or scaling up/down Kafka, there are no errors in the log of CC.

### Checklist

- [ ] Make sure all tests pass
- [ ] Update documentation

### CODE REVIEW

File: systemtest/src/test/java/io/strimzi/systemtest/AbstractClusterIT.java [hunk 1]
```diff
@@ -168,4 +170,10 @@ public String consumeMessages(String clusterName, String topic, int groupID, int
                             "-kafka:9092 --topic " + topic + " --group-id " + groupID + " & sleep "
                             + timeout + "; kill %1").out();
     }
+
+    void checkErrorsInLogCC() {
+        //TODO add blacklist for unexpected errors
+        String searchPattern = "\'Exception\\|Error\\|Throwable\'";
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/AbstractClusterIT.java]: The `'` quotes belong in `searchInLog()` – I don't think callers of the method should care that you happen to be passing the regex to grap via `bash -c` and therefore need to be quoted for the shell: It's an implementation detail of `searchInLog()`.  I suppose the same applies to the need to `\`-quote the `|`, but TBH it's not clear to me whether it's bash or grep which is handling that quoting – do you know?

File: systemtest/src/test/java/io/strimzi/systemtest/AbstractClusterIT.java [hunk 2]
```diff
@@ -168,4 +170,10 @@ public String consumeMessages(String clusterName, String topic, int groupID, int
                             "-kafka:9092 --topic " + topic + " --group-id " + groupID + " & sleep "
                             + timeout + "; kill %1").out();
     }
+
+    void checkErrorsInLogCC() {
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/AbstractClusterIT.java]: `assertNoCcErrorsLogged()` might be a better name – it is ultimately an assertion.

File: systemtest/src/test/java/io/strimzi/systemtest/AbstractClusterIT.java (hunk 1 above)
Comment (serrss) [reply systemtest/src/test/java/io/strimzi/systemtest/AbstractClusterIT.java]: @tombentley TBH I tried to use many patterns (without `'` quotes, without `\|`) but in practice it turns out that if we want to use many keywords for executing `grep` through `Exec` we need to use all added symbols and it was a reason of using search pattern for it. But in another hand (as you said) the callers don't care about `'` quotes etc, so looks like the better way will be move regex logic to method `searchInLog()` if I understood you right

File: common-test/src/main/java/io/strimzi/test/k8s/BaseKubeClient.java [hunk 3]
```diff
@@ -354,4 +355,19 @@ public String describe(String resourceType, String resourceName) {
     public String logs(String pod) {
         return Exec.exec(namespacedCommand("logs", pod)).out();
     }
+
+    @Override
+    public String searchInLog(String resourceType, String resourceName, String sinceSeconds, String... grepPattern) {
+        try {
+            return Exec.exec("bash", "-c", join(" ", namespacedCommand("logs", resourceType + "/" + resourceName, "--since=" + sinceSeconds + "s",
+                    "|", "grep", "\'" + join("\\|", grepPattern) + "\'"))).out();
```
Comment (tombentley) [inline common-test/src/main/java/io/strimzi/test/k8s/BaseKubeClient.java]: The reason you're having to quote the `|` is `grep` defaults to using basic REs. You could use `grep -E`, which should mean you don't need to quite the pipes. Or you could use multiple `-e` options to save having to combine the patterns into a single one.

**Pull request merged**